### ISSUE TITLE
extension.all API returns outdated result when onDidGrantWorkspaceTrust 

### ISSUE THREAD

**Issue by Eskibear** [microsoft/vscode#127055]
- VS Code Version: 1.58-insiders
- OS Version: Windows 10

Context information:

Java extension has a contribution point for Java LS. Other extensions (e.g. Java debugger) can declare plugins for Java LS in their package.json files. Before starting Java LS, it calls `vscode.extensions.all` to collect plugin information and load extra libs. 

Problem:

When you switching workspace from untrusted to trusted, at the time `onDidGrantWorkspaceTrust` is triggered, result of `extension.all` doesn't contain those extensions with `untrustedWorkspace: false`. So in our case, Java debugger extension is activated eventually, but the libs declared in its package.json are not loaded by Java extension.  (reported in https://github.com/redhat-developer/vscode-java/issues/1994)

Steps to Reproduce:

1. create an extension, with `limited` untrustedWorkspaces support
2. add listener `onDidGrantWorkspaceTrust(() => console.log(vscode.extensions.all));`
3. open an untrusted workspace, then trust it, inspect console log.

Currently our workaround can be either 
a) add timeout with a magic number, like 1000ms. [sample commit](https://github.com/redhat-developer/vscode-java/pull/1995/commits/abdd035e32a72acaa9082107615b7c5d19e5de34#diff-04bba6a35cad1c794cbbe677678a51de13441b7a6ee8592b7b50be1f05c6f626R355-R360)
b) maybe work together with API `extensions.onDidChange()`?

Since the name is "on**Did**GrantWorkspaceTrust" instead of on**Will**GrantWorkspaceTrust, I intuitively think `extension.all` should return a full list of the extensions at that time. 

@lszomoru What do you think?

**Comment by testforstephen** [comment]
> b) maybe work together with API `extensions.onDidChange()`?

To clarify, the workaround `b` doesn't work. `extensions.onDidChange()` is not called when granting workspace trust.

**Comment by lszomoru** [comment]
@Eskibear, we have recently done some refactoring around extension enablement that is only preset in Insiders.
Are you seeing the same behaviour in Stable as well, or is this behaviour limited to Insiders only? Thanks!

**Comment by Eskibear** [comment]
> Are you seeing the same behaviour in Stable as well

Yes, I can reproduce this issue in stable 1.57.1. 

In below video, I add some statements and breakpoints to inspect `extensions.all`.
- In untrusted state, the first time I triggered the command, length of `extensions.all` was 88.
- onDidGrantWorkspaceTrust, `extensions.all` still returned 88.
- I triggered the command to inspect `extension.all` again, the length was 94.


https://user-images.githubusercontent.com/2351748/123271100-12c6d100-d533-11eb-932f-1e267c229068.mp4

**Comment by alexdima** [comment]
@Eskibear `vscode.extensions.all` has its own dedicated event, `vscode.extensions.onDidChange`. Extensions can be added or removed for numerous reasons (e.g. installing, enabling, disabling, uninstalling, etc.), so any logic that needs to deal with `vscode.extensions.all` should listen to  `vscode.extensions.onDidChange` and not to some other event.

**Comment by leodevbro** [comment]
@alexdima, well, in complex code I somehow remember it returned undefined sometimes, unfortunately I don't remember how to reproduce it. But the main problem is that it returns only one extension (the current building extension), it never returns all the extensions. This is for `yo code -> New Web Extension (TypeScript)`, but for `yo code -> New Extension (TypeScript)` it returns all extensions.

![2022-03-03_04-06-45](https://user-images.githubusercontent.com/54504880/156471267-8d749bc1-0128-44de-bd5e-5ccb373450b2.png)

**Comment by leodevbro** [comment]
@alexdima, thanks, I tried and it really works at `https://vscode.dev` environment. `vscode.extensions.all` gets all the extensions.
Now I got another issue about the web VSCode environment. I do have access to the extensions but it seems I cannot read files of them. Well, I guess this is the last stop for me struggling to port Blockman into web VSCode, I guess it is the API limitation and maybe we cannot do anything about it, right? **Anyway, thank you so much for your support.**

![vscode-web-extension-issue007](https://user-images.githubusercontent.com/54504880/157560253-c49096e6-cc7f-4b13-8eb3-4ea31e3b884a.png)

**Comment by alexdima** [comment]
@leodevbro In your above snippet, where does `configPath` come from? Each extension (`Extension<T>`) has a property called `extensionUri` and you should use something like `vscode.Uri.joinPath(extension.extensionUri, pathInsideExtension)` to get a valid `Uri`; the assumption that this is a `file://` URI doesn't hold on the web.

**Comment by leodevbro** [comment]
@alexdima, Well, I tried your instructions now. please see the parts of original and altered code here:

```typescript
for (const language of extension.packageJSON.contributes.languages) {
    if (language.configuration) {
  
   // this is the original code line which works fine in desktop (node) environment:
      this.languageToConfigPath.set(language.id, path.join(extension.extensionPath, language.configuration)); 

   // this is how I'm trying to alter it like this:
      this.languageToConfigUri.set(language.id, vscode.Uri.joinPath(extension.extensionUri, extension.extensionUri.path));
    }
  }

//
const languageID: any;



// this is the original code line to get the path:
const configPath = this.languageToConfigPath.get(languageID);

// this is my code line to get the Uri:
const configUri = this.languageToConfigUri.get(languageID);
```

**My altered code seems not working when trying to read file of the URI:**

![image](https://user-images.githubusercontent.com/54504880/157967831-1206e21f-3b4d-4ba9-a9dd-83904daa1d4c.png)

Maybe this:
`vscode.Uri.joinPath(extension.extensionUri, extension.extensionUri.path)`
is not the correct alteration for this:
`path.join(extension.extensionPath, language.configuration)`
right?

Could you please give me a bit more instructions for it?


Here, this is just one of many extensions console.logged just to see how it looks like in my the web browser:

![image](https://user-images.githubusercontent.com/54504880/157968264-c4f56c3b-5cea-425f-803d-0d762434cbc5.png)

**Comment by alexdima** [comment]
The equivalent of:
```
path.join(extension.extensionPath, language.configuration)
```
would be:
```
vscode.Uri.joinPath(extension.extensionUri, language.configuration)
```

**Comment by leodevbro** [comment]
@alexdima, Thanks, it seems working. Now I got another strange issue. **Please feel free to ignore this new issue**, because I guess it requires a lot of time to investigate.

I'll post it here just in case. And also I'm still trying to investigate it. Maybe I will be able to figure it out in several days.

This is the full code of the entire file:
https://gist.github.com/leodevbro/a8e9b9f4a6c8545d9ca217bb77704606

The error is here, and I am 95% sure that it is caught inside `loadGrammarWithConfiguration` function.
**Important note:** I tested this file inside desktop (node) environmental extension project and it catches absolutely the same error.

![2022-03-16_15-17-54--](https://user-images.githubusercontent.com/54504880/158582655-bf98140f-d130-41e4-b3e3-45d921c28555.png)
